features/323-login-id/design.md

```diff
@@ -0,0 +1,165 @@
+login ID component, since the format of serialization is left unspecified.
+
+To support compound login ID better, a Realm would be associated with each
+password authenticated principal.
+- Realm is a user specified value, auth gear does not concern its semantics.
+- Default realm is 'default' if unspecified, for application without need of compound login ID.
+- Possible realm values are whitelisted.
+- Realm would be considered when matching login ID with principal.
+  i.e. password authenticated principals has unique (realm, login ID).
+- API consumer could specify realm along with login ID when realm is required.
+
+This solution should support majority of use case for compound login ID.
+
+For example, when logging in with realms:
+```json
+{ "realm": "teacher", "login_id": "a@example.com", "password": "12345678" }
+{ "realm": "student", "login_id": "a@example.com", "password": "12345678" }
+```
+If authentication succeed, both request would result in same user, with
+different principals.
+
+This addresses use case 10.
+
+## Allow Multiple Login ID per Login ID Key
+In some website, user can associate multiple emails (potentially unlimited) with
+same user. These email may be used in auth features:
+- Login with multiple emails
+- Backup email
+- Welcome email
+
+A login ID key should allow having multiple associated login ID.
+When signing up, user should provide a list of login ID. e.g.
+```json
+{
+    "login_ids": [
+        { "username": "kiootic" },
+        { "email": "kiootic+1@oursky.com" },
+        { "email": "kiootic+2@oursky.com" }
+    ],
+    "password": "12345678"
+}
+```
+
+This addresses use case 6.
+
+
+## Standard Keys
+With reference to OpenID Connect standard claims, a list of standard keys
+would be specified:
+- `username`
+- `email`
+- `phone`
+- etc.
+
+These keys would be used in user metadata.
+
+
+## Configurable Login ID Keys
+The login ID keys are whitelisted by developer. In additional, developer can
+configure each login ID key:
+- The corresponding standard key (normalization may be applied)
+- The minimum/maximum amount of login ID
+
+The auth gear would use configured keys in password recovery and welcome message.
+
+
+## User Settings
+User settings would be added. This settings is a map of strings, and only keys
+recognized by auth gear is allowed:
+- `primary_login_id`
+
+The settings have default values, and its value can be updated later using API.
+
+This settings is intended to be consumed by auth features:
+- Welcome email/SMS: Can be configured to send to primary login ID/all supported login ID.
+
+# Design Detail
+This specification would be implemented in stages:
+1. [Implicit login ID key, Login ID realm](stage-1.md)
+2. [Multiple login IDs per key, Configurable login ID keys, Standard keys](stage-2.md)
+3. [User settings](stage-3.md)
+
+Design detail of each stage can be found in corresponding documents.
+
+# Appendix
```

Comment:
@kiootic I think it will help future us to understand better, if we can include a sample of user object at appendix / somewhere in this document, which include all parameters of a user object, metadata sample, and the new user settings.

Reply:
Especially it feels a bit confusing on how verified/disabled would work after the changes above. And which attributes go to User Setting, which goes to the user object? (i.e. why verified is in user object, while primary_login_id goes to user setting?)

Reply:
I'll add another document to summarize the new APIs and logic, with sample use cases.

Reply:
Thanks. Ideally it is in the main doc as that should be the main thing we're designing.

(In fact, features spec DO NOT always need to discuss how to implement, we mainly want to discuss how users will use it...)

Reply:
Notes from the other thread: please include createLoginID / deleteLoginID in it.